// datocms-client/lib/local/EntitiesRepo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _JsonApiEntity = _interopRequireDefault(require("./JsonApiEntity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var entitiesToStorePerCacheKey = 5000;

function payloadEntities(payload) {
  var accumulator = [];

  if (payload.data) {
    if (Array.isArray(payload.data)) {
      accumulator = [].concat(_toConsumableArray(accumulator), _toConsumableArray(payload.data));
    } else {
      accumulator.push(payload.data);
    }
  }

  if (payload.included) {
    accumulator = [].concat(_toConsumableArray(accumulator), _toConsumableArray(payload.included));
  }

  return accumulator;
}

var EntitiesRepo = /*#__PURE__*/function () {
  function EntitiesRepo() {
    _classCallCheck(this, EntitiesRepo);

    this.entities = {};
    this.destroyListeners = [];
    this.upsertListeners = [];
    this.upsertEntities.apply(this, arguments);
  }

  _createClass(EntitiesRepo, [{
    key: "saveStateToCache",
    value: function () {
      var _saveStateToCache = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(cache, cachePrefixKey) {
        var entityTypes, manifest, _i, _entityTypes, entityType, entities, i, chunkIndex, chunkCacheKey;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                entityTypes = Object.keys(this.entities);
                manifest = {
                  entityTypeChunkKeys: {}
                };
                _i = 0, _entityTypes = entityTypes;

              case 3:
                if (!(_i < _entityTypes.length)) {
                  _context.next = 19;
                  break;
                }

                entityType = _entityTypes[_i];
                entities = Object.values(this.entities[entityType]);
                i = 0, chunkIndex = 0;

              case 7:
                if (!(i < entities.length)) {
                  _context.next = 16;
                  break;
                }

                chunkCacheKey = "".concat(cachePrefixKey, "--").concat(entityType, "-").concat(chunkIndex);
                manifest.entityTypeChunkKeys[entityType] = manifest.entityTypeChunkKeys[entityType] || [];
                manifest.entityTypeChunkKeys[entityType].push(chunkCacheKey);
                _context.next = 13;
                return cache.set(chunkCacheKey, entities.slice(i, i + entitiesToStorePerCacheKey).map(function (entity) {
                  return entity.payload;
                }));

              case 13:
                i += entitiesToStorePerCacheKey, chunkIndex += 1;
                _context.next = 7;
                break;

              case 16:
                _i++;
                _context.next = 3;
                break;

              case 19:
                _context.next = 21;
                return cache.set(cachePrefixKey, manifest);

              case 21:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function saveStateToCache(_x, _x2) {
        return _saveStateToCache.apply(this, arguments);
      }

      return saveStateToCache;
    }()
  }, {
    key: "loadStateFromCache",
    value: function () {
      var _loadStateFromCache = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(cache, cachePrefixKey) {
        var _this = this;

        var manifest, _loop, _i2, _Object$entries;

        return regeneratorRuntime.wrap(function _callee2$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return cache.get(cachePrefixKey);

              case 2:
                manifest = _context3.sent;

                if (manifest) {
                  _context3.next = 5;
                  break;
                }

                return _context3.abrupt("return");

              case 5:
                this.entities = {};
                _loop = /*#__PURE__*/regeneratorRuntime.mark(function _loop() {
                  var _Object$entries$_i, entityType, entityTypeChunkKeys, _iterator, _step, entityTypeChunkKey, chunkEntities;

                  return regeneratorRuntime.wrap(function _loop$(_context2) {
                    while (1) {
                      switch (_context2.prev = _context2.next) {
                        case 0:
                          _Object$entries$_i = _slicedToArray(_Object$entries[_i2], 2), entityType = _Object$entries$_i[0], entityTypeChunkKeys = _Object$entries$_i[1];
                          _this.entities[entityType] = {};
                          _iterator = _createForOfIteratorHelper(entityTypeChunkKeys);
                          _context2.prev = 3;

                          _iterator.s();

                        case 5:
                          if ((_step = _iterator.n()).done) {
                            _context2.next = 13;
                            break;
                          }

                          entityTypeChunkKey = _step.value;
                          _context2.next = 9;
                          return cache.get(entityTypeChunkKey);

                        case 9:
                          chunkEntities = _context2.sent;
                          chunkEntities.forEach(function (entityPayload) {
                            _this.entities[entityType][entityPayload.id] = new _JsonApiEntity["default"](entityPayload, _this);
                          });

                        case 11:
                          _context2.next = 5;
                          break;

                        case 13:
                          _context2.next = 18;
                          break;

                        case 15:
                          _context2.prev = 15;
                          _context2.t0 = _context2["catch"](3);

                          _iterator.e(_context2.t0);

                        case 18:
                          _context2.prev = 18;

                          _iterator.f();

                          return _context2.finish(18);

                        case 21:
                        case "end":
                          return _context2.stop();
                      }
                    }
                  }, _loop, null, [[3, 15, 18, 21]]);
                });
                _i2 = 0, _Object$entries = Object.entries(manifest.entityTypeChunkKeys);

              case 8:
                if (!(_i2 < _Object$entries.length)) {
                  _context3.next = 13;
                  break;
                }

                return _context3.delegateYield(_loop(), "t0", 10);

              case 10:
                _i2++;
                _context3.next = 8;
                break;

              case 13:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee2, this);
      }));

      function loadStateFromCache(_x3, _x4) {
        return _loadStateFromCache.apply(this, arguments);
      }

      return loadStateFromCache;
    }()
  }, {
    key: "serializeState",
    value: function serializeState() {
      return Object.entries(this.entities).reduce(function (acc, _ref) {
        var _ref2 = _slicedToArray(_ref, 2),
            type = _ref2[0],
            entitiesById = _ref2[1];

        return _objectSpread(_objectSpread({}, acc), {}, _defineProperty({}, type, Object.entries(entitiesById).reduce(function (acc2, _ref3) {
          var _ref4 = _slicedToArray(_ref3, 2),
              id = _ref4[0],
              entity = _ref4[1];

          return _objectSpread(_objectSpread({}, acc2), {}, _defineProperty({}, id, entity.payload));
        }, {})));
      }, {});
    }
  }, {
    key: "loadState",
    value: function loadState(serializedState) {
      var _this2 = this;

      this.entities = Object.entries(serializedState).reduce(function (acc, _ref5) {
        var _ref6 = _slicedToArray(_ref5, 2),
            type = _ref6[0],
            entitiesById = _ref6[1];

        return _objectSpread(_objectSpread({}, acc), {}, _defineProperty({}, type, Object.entries(entitiesById).reduce(function (acc2, _ref7) {
          var _ref8 = _slicedToArray(_ref7, 2),
              id = _ref8[0],
              payload = _ref8[1];

          return _objectSpread(_objectSpread({}, acc2), {}, _defineProperty({}, id, new _JsonApiEntity["default"](payload, _this2)));
        }, {})));
      }, {});
    }
  }, {
    key: "addDestroyListener",
    value: function addDestroyListener(cb) {
      var _this3 = this;

      this.destroyListeners.push(cb);
      return function () {
        _this3.destroyListeners = _this3.destroyListeners.filter(function (x) {
          return x !== cb;
        });
      };
    }
  }, {
    key: "addUpsertListener",
    value: function addUpsertListener(cb) {
      var _this4 = this;

      this.upsertListeners.push(cb);
      return function () {
        _this4.upsertListeners = _this4.upsertListeners.filter(function (x) {
          return x !== cb;
        });
      };
    }
  }, {
    key: "destroyAllEntities",
    value: function destroyAllEntities() {
      var _this5 = this;

      // Order is important here! See ie. gatsby-source-datocms `destroyEntityNode` function!
      ['item', 'field', 'item_type', 'upload', 'site'].forEach(function (type) {
        if (_this5.entities[type]) {
          _this5.destroyEntities(type, Object.keys(_this5.entities[type]));
        }
      });
    }
  }, {
    key: "upsertEntities",
    value: function upsertEntities() {
      var _this6 = this;

      var entities = [];

      for (var _len = arguments.length, payloads = new Array(_len), _key = 0; _key < _len; _key++) {
        payloads[_key] = arguments[_key];
      }

      payloads.forEach(function (payload) {
        payloadEntities(payload).forEach(function (entityPayload) {
          var entity = new _JsonApiEntity["default"](entityPayload, _this6);
          entities.push(entity);
          _this6.entities[entity.type] = _this6.entities[entity.type] || {};
          _this6.entities[entity.type][entity.id] = entity;
        });
      });
      this.upsertListeners.forEach(function (cb) {
        return entities.forEach(cb);
      });
    }
  }, {
    key: "destroyEntities",
    value: function destroyEntities(type, ids) {
      var _this7 = this;

      var entities = [];
      ids.forEach(function (id) {
        if (_this7.entities[type] && _this7.entities[type][id]) {
          entities.push(_this7.entities[type][id]);
          delete _this7.entities[type][id];
        }
      });
      this.destroyListeners.forEach(function (cb) {
        return entities.forEach(cb);
      });
    }
  }, {
    key: "destroyItemType",
    value: function destroyItemType(id) {
      var itemIds = Object.values(this.entities.item).filter(function (item) {
        return item.itemType.id === id;
      }).map(function (item) {
        return item.id;
      });
      this.entitiesRepo.destroyEntities('item', itemIds);
      this.entitiesRepo.destroyEntities('item_type', [id]);
    }
  }, {
    key: "findEntitiesOfType",
    value: function findEntitiesOfType(type) {
      return Object.values(this.entities[type] || {});
    }
  }, {
    key: "findEntity",
    value: function findEntity(type, id) {
      return this.entities[type] && this.entities[type][id];
    }
  }, {
    key: "site",
    get: function get() {
      return this.findEntitiesOfType('site')[0];
    }
  }]);

  return EntitiesRepo;
}();

exports["default"] = EntitiesRepo;
